troposphere/haze/source/ptp_responder.cpp

```diff
@@ -866,7 +884,82 @@ namespace haze {
+        R_TRY(dp.Read(std::addressof(data_header)));
+        R_UNLESS(data_header.type == PtpUsbBulkContainerType_Data,      haze::ResultUnknownRequestType());
+        R_UNLESS(data_header.code == m_request_header.code,             haze::ResultOperationNotSupported());
+        R_UNLESS(data_header.trans_id == m_request_header.trans_id,     haze::ResultOperationNotSupported());
+
+        /* Ensure we have a valid property code before continuing. */
+        R_UNLESS(property_code == PtpObjectPropertyCode_ObjectFileName, haze::ResultUnknownPropertyCode());
+
+        /* Check if we know about the object. If we don't, it's an error. */
+        auto * const obj = m_object_database.GetObjectById(object_id);
+        R_UNLESS(obj != nullptr, haze::ResultInvalidObjectId());
+
+        /* We are reading a file name. */
+        R_TRY(dp.ReadString(g_filename_str));
+        R_TRY(dp.Finalize());
+
+        /* Ensure we can actually process the new name. */
+        const bool is_empty         = g_filename_str[0] == '\x00';
+        const bool contains_slashes = std::strchr(g_filename_str, '/') != nullptr;
+        R_UNLESS(!is_empty && !contains_slashes, haze::ResultInvalidPropertyValue());
+
+        /* Add a new object in the database with the new name. */
+        PtpObject *newobj;
+        {
+            /* Find the last path separator in the existing object name. */
+            char *pathsep = std::strrchr(obj->m_name, '/');
+            HAZE_ASSERT(pathsep != nullptr);
+
+            /* Temporarily mark the path separator as null to facilitate processing. */
+            *pathsep = '\x00';
+            ON_SCOPE_EXIT { *pathsep = '/'; };
+
+            R_TRY(m_object_database.CreateOrFindObject(obj->GetName(), g_filename_str, obj->GetParentId(), std::addressof(newobj)));
+        }
+
+        {
+            /* Ensure we maintain a clean state on failure. */
+            ON_RESULT_FAILURE {
+                if (!newobj->GetIsRegistered()) {
+                    /* Only delete if the object was not registered. */
+                    /* Otherwise, we would remove an object that still exists. */
+                    m_object_database.DeleteObject(newobj);
+                }
+            };
+
+            /* Get the old object type. */
+            FsDirEntryType entry_type;
+            R_TRY(m_fs.GetEntryType(obj->GetName(), std::addressof(entry_type)));
+
+            /* Attempt to rename the object on the filesystem. */
+            if (entry_type == FsDirEntryType_Dir) {
+                R_TRY(m_fs.RenameDirectory(obj->GetName(), newobj->GetName()));
+            } else {
+                R_TRY(m_fs.RenameFile(obj->GetName(), newobj->GetName()));
+            }
+        }
+
+        /* Unregister and free the old object. */
+        m_object_database.DeleteObject(obj);
+
+        /* Register the new object. */
+        m_object_database.RegisterObject(newobj, object_id);
+
+        /* We succeeded. */
```

Comment:
Typically, "We succeeded" comments should only be put above an R_SUCCEED, because you have R_TRY'd the failable calls. This comment is a little strange to me, though it makes sense -- you are writing an "Ok" response code. You do not need to change this, and the comment appears elsewhere, just noting my confusion for posterity.